# Importing libraries

In [40]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [41]:
#creating path

pathk=r'/Users/keerthiputti/Desktop/Data Analysis/Instacart Basket Analysis-Python'

In [42]:
print(pathk)

/Users/keerthiputti/Desktop/Data Analysis/Instacart Basket Analysis-Python


In [17]:
# ctreating dataframe for departments

df_dep=pd.read_csv(os.path.join(pathk,'02 Data','Original Data','departments.csv'),index_col=False)

In [19]:
#transposing

df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [21]:
df_dep_t=df_dep.T

In [22]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
#restting index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
#new header
new_header=df_dep_t.iloc[0]

In [25]:
new_header

0    department
Name: department_id, dtype: object

In [26]:
df_dep_new=df_dep_t[1:]

In [27]:
df_dep_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [28]:
df_dep_new.columns=new_header

In [29]:
df_dep_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [31]:
#create data frames for orders
df_ords=pd.read_csv(os.path.join(pathk,'02 Data','Original Data','Orders.csv'),index_col=False)

In [32]:
#create data frame for products
df_prod=pd.read_csv(os.path.join(pathk,'02 Data','Original Data','products.csv'),index_col=False)

In [33]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [34]:
df_prod.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [63]:
#data dictionary for depratement DF
data_dict=df_dep_new.to_dict('index')

In [66]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [39]:
df_prod.shape

(49693, 5)

# Begining of task 4.4, libraries are imported at the top and dataframes created for order and products

In [43]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [46]:
#dropping column eval_set
df_ords_dr=df_ords.drop(columns=['eval_set'])

In [47]:
df_ords_dr

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [48]:
# 4a. changing data type for order_id and user_id
df_ords_dr['order_id']=df_ords_dr['order_id'].astype('str')

In [49]:
df_ords_dr['user_id']=df_ords_dr['user_id'].astype('str')

In [55]:
#checking datatype change
df_ords_dr.dtypes

order_id                   object
user_id                    object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [56]:
#4b renaming order_dow column
df_ords_dr.rename(columns={"order_dow":"Order_day_of_the_week"},inplace=True)

In [57]:
df_ords_dr.head()

order_id user_id  order_number  Order_day_of_the_week  order_hour_of_day  \
0  2539329       1             1                      2                  8   
1  2398795       1             2                      3                  7   
2   473747       1             3                      3                 12   
3  2254736       1             4                      4                  7   
4   431534       1             5                      4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [58]:
#4c finding the highest orders in a day
df_ords_dr['order_hour_of_day'].value_counts(dropna=False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

looking at the frequency for order_hour_of_day, we can understand Morning 10 AM is the busiest time.

In [68]:
# 4d data_dict is the dictionary for departement. we will use it here to understand what is department 4 is
print(data_dict.get('4'))

{'department': 'produce'}


In [69]:
#4e checking for breakfast items and creating subset
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

Breakfast item is in department 14. We will use this to create subset

In [71]:
#breakfast item subset creation
df_brfst_sub=df_prod.loc[df_prod['department_id']==14]

In [72]:
df_brfst_sub

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [73]:
#4f creating subset for party items
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

item- alchohol,deli,beverages,meat/seafood - departement_ids 5,7,12 and 20

In [78]:
#creating subset of party items
df_party_sub=df_prod.loc[df_prod['department_id'].isin([5,7,12,20])]

In [80]:
df_party_sub

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

4g Last created dataframe is df_party_sub has 7650 rows. this can be checked using df_party_sub.shape as well.

In [83]:
#4h creating subset for user_id 1
df_userid1_sub=df_ords_dr.loc[df_ords_dr['user_id']=='1']

In [84]:
#all information about user_id 1
df_userid1_sub

order_id user_id  order_number  Order_day_of_the_week  order_hour_of_day  \
0   2539329       1             1                      2                  8   
1   2398795       1             2                      3                  7   
2    473747       1             3                      3                 12   
3   2254736       1             4                      4                  7   
4    431534       1             5                      4                 15   
5   3367565       1             6                      2                  7   
6    550135       1             7                      1                  9   
7   3108588       1             8                      1                 14   
8   2295261       1             9                      1                 16   
9   2550362       1            10                      4                  8   
10  1187899       1            11                      4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [86]:
#4i extracting stats about user_id 1
df_userid1_sub.describe()

order_number  Order_day_of_the_week  order_hour_of_day  \
count     11.000000              11.000000          11.000000   
mean       6.000000               2.636364          10.090909   
std        3.316625               1.286291           3.477198   
min        1.000000               1.000000           7.000000   
25%        3.500000               1.500000           7.500000   
50%        6.000000               3.000000           8.000000   
75%        8.500000               4.000000          13.000000   
max       11.000000               4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [87]:
#4j exporting orders wrangled data
df_ords_dr.to_csv(os.path.join(pathk,'02 Data','Prepared Data','Orders_wrangled.csv'))

In [88]:
#4k Exporting departement data frame
df_dep_new.to_csv(os.path.join(pathk,'02 Data','Prepared Data','Department_wrangled.csv'))